### DISCUSSION TITLE
Asynchronous calls in custom nodes

### DISCUSSION THREAD

**Discussion by HonzaMac** [markdoc/markdoc#90] [Q&A] [answered]
We are using Markdoc for parsing `md` file with images. We need to upload `md` file to API together with images.

We tried to write plugin for parsing images which would upload file to external API and replace current `src` with new location of file.
Unfortunately we failed with this approach as we found, that `transform` method can't be `async`.

Is there some recommended way to write custom "node" with async code?

We had this: 
```js
export const imageUpload = (): Schema => ({
  render: 'img',
  attributes: {
    src: {type: String, required: true},
    alt: {type: String},
  },
  async transform(node, config) {
    const attributes = node.transformAttributes(config) as {src: string; alt?: string};
    assertIsDefined(attributes?.src, Error, 'src have to be defined');

    const location = await uploadImage(attributes.src);
    return new Tag('img', {...attributes, src: location}, node.transformChildren(config));
  },
});
```

**Comment by mfix-stripe** [comment, marked as answer]
Hey @HonzaMac, good question! We have heard similar questions in the past, and I could see it being something we might support in the future. 

For your particular use case, though, I don't think you should be using the `transform` function to upload the images. You should upload the images outside of the Markdoc lifecycle, and "look up" the image source in the `transform` function.

**Reply by mfix-stripe** [reply]
On another note, I see you added `assertIsDefined` to your `transform` function—you should use `validate` for this! We have docs on how you might do this: https://markdoc.io/docs/validation#validating-attributes

**Reply by ra100** [reply]
I can see that having transform async could help with some transformations and make them much simpler compared to parse, do something async, have some matcher or what and continue with transformation.

Some content generators can be async, I'd say, even loading some file can be prefered to be async rather than use sync functions of `fs`. Or if you want to use remote service to generate/transform some content for you. Or just load content from remote page, for example from confluence.

**Comment by nick-kang** [comment]
Another use case for async transform is that [next/image](https://nextjs.org/docs/api-reference/next/image) requires `width` and `height` (except for `layout=fill`). We'd like to asynchronously fetch the image and determine its width and height so we can pass it to the component.

The alternative is to have the user manually input the width/height, which can be error prone.

**Comment by HonzaMac** [comment]
Thank you for recent release of async [experimental feature in 0.1.4](https://github.com/markdoc/markdoc/releases/tag/0.1.4).

We have tried it and it works. We could rapidly simplify our code for uploading images.

I think that `validate(node, config)` should be also `async`. 
For use-case of image upload, we can validate existence of file in validation. User would then get validation error on non-existent file instead of waiting for transformation and stop uploading of many files in the middle.

What we currently **can't do**: 
```typescript
import {stat} from 'fs/promises';

export const replaceImagesOnTransform = (accessToken: string, directory: string): Schema => ({
  render: TAG_NAME,
  attributes: {
    src: {type: String, required: true},
    alt: {type: String},
  },
  async validate(node, config) {
    const {src} = node.transformAttributes(config);
    try {
      await stat(src, {throwIfNoEntry: true});
    } catch (e) {
      return [{id: 'file-not-found', message: `File "${src}" not found. ${e?.message}`, level: 'error'}];
    }
  },
  async transform(node, config) {
    const attributes = node.transformAttributes(config);
    const location = getLocation(attributes.src, directory);
    const src = await uploadSingleFile(location, accessToken);

    return new Tag(TAG_NAME, {...attributes, src}, node.transformChildren(config));
  },
});
```

Thank you.
